pdns/dnsparser.cc

```diff
@@ -553,6 +553,107 @@ void PacketReader::xfrBlob(string& blob, int length)
   }
 }
 
+void PacketReader::xfrSvcParamKeyVals(set<SvcParam> &kvs) {
+  while (d_pos < (d_startrecordpos + d_recordlen)) {
+    if (d_pos + 2 > (d_startrecordpos + d_recordlen)) {
+      throw std::out_of_range("incomplete key");
+    }
+    uint16_t keyInt;
+    xfr16BitInt(keyInt);
+    auto key = static_cast<SvcParam::SvcParamKey>(keyInt);
+    uint16_t len;
+    xfr16BitInt(len);
+    
+    if (d_pos + len > (d_startrecordpos + d_recordlen)) {
+      throw std::out_of_range("record is shorter than SVCB lengthfield implies");
+    }
+
+    switch (key)
+    {
+    case SvcParam::mandatory: {
+      if (len % 2 != 0) {
+        throw std::out_of_range("mandatory SvcParam has invalid length");
+      }
+      if (len == 0) {
+        throw std::out_of_range("empty 'mandatory' values");
+      }
+      std::set<SvcParam::SvcParamKey> paramKeys;
+      size_t stop = d_pos + len;
+      while (d_pos < stop) {
+        uint16_t keyval;
+        xfr16BitInt(keyval);
+        paramKeys.insert(static_cast<SvcParam::SvcParamKey>(keyval));
+      }
+      kvs.insert(SvcParam(key, paramKeys));
```

Comment:
It would be a nice improvement to make the constructors move the second parameter, instead of doing copies.